spec/xml_security_spec.rb

```diff
@@ -94,6 +94,37 @@ module SamlIdp
         end
       end
 
+      describe '#digest_method_algorithm' do
+        let(:document) do
+          XMLSecurity::SignedDocument.new(fixture(:valid_sha256_no_ds_namespace, false))
+        end
+        let(:sig_namespace_hash) { { 'ds' => 'http://www.w3.org/2000/09/xmldsig#' } }
+
+        let(:el) do
+          REXML::XPath.first(
+            document,
+            '//ds:Signature',
+            sig_namespace_hash
+          )
+        end
+
+        let(:sig_element) do
+          REXML::XPath.first(el, '//ds:Reference', sig_namespace_hash)
+        end
+
+        let(:ref) do
+          REXML::XPath.first(sig_element, '//ds:Reference', sig_namespace_hash)
+        end
+
+        it 'returns the value in the DigestMethod node' do
+          expect(document.send(
+            :digest_method_algorithm,
+            ref,
+            sig_namespace_hash
+          )).to eq OpenSSL::Digest::SHA256
+        end
+      end
```

Comment:
do we need a case of passing in wrong `sig_namespace_hash` and getting no result?

Reply:
i think currently no because of another possible subtle bug i need to investigate. as far as i can tell, if we get to this part of the code, the `sig_namespace_hash` will always be `{ 'ds' => 'http://www.w3.org/2000/09/xmldsig#' }`. 

https://github.com/18F/saml_idp/blob/main/lib/saml_idp/xml_security.rb#L168-L174

Reply:
clarification: i think it evaluates to nil if there is no `Signature` element, but if there's no `Signature` element, it doesn't hit this code path.